diff --git a/views_bulk_operations.rules.inc b/views_bulk_operations.rules.inc
index e664ee6..da841d9 100644
--- a/views_bulk_operations.rules.inc
+++ b/views_bulk_operations.rules.inc
@@ -113,7 +113,7 @@ function views_bulk_operations_rules_action_info() {
  */
 function views_bulk_operations_views_list() {
   $selectable_displays = array();
-  foreach (views_get_all_views() as $name => $view) {
+  foreach (views_get_enabled_views() as $name => $view) {
     foreach ($view->display as $display_name => $display) {
       $view->build($display_name);
       $vbo = _views_bulk_operations_get_field($view);
